### PR TITLE
Scala 3 support

### PR DESCRIPTION
**Pull request by grzegorz-bielski**
TODO:
- [x] Fixing exercises
- [x] New test-gen POC
- [x] Scala 3 test runner: https://github.com/exercism/scala-test-runner/pull/55

Forum thread: https://forum.exercism.org/t/scala-3-support/12117

Supersedes: https://github.com/exercism/scala/pull/734

### CODE REVIEW

**Comment by grzegorz-bielski** [comment]
I tried to port test generators to Scala 3, but looks like it's not going to be that easy.

I was able to update the libraries to the newest versions, but their up-to-date version simply doesn't work with the unsafe hack that was used in the test generators where `JsValue` from `play-json` was casted to `Map[String, Any]` . 
In order to make test generators working again they have to be refactored at the ground level and each individual test generator should be ported, one by one. 
It's not that hard, but requires a significant time commitment for which I'm not ready a the moment.
I already made a new POC setup based on the old one and ported one test generator for the `word-count` exercise.

My current development plan:
- for now I'll focus on porting only the `exercises` folder, just to make it compile with Scala 3. Even if something was auto-generated before, I won't touchits test-gen code. This way we can move forward, and people could start using Scala 3 in the exercism.
- I'll remove the old test-gen setup and create a new one, where we can port test-generators one by one.
From what I can see at: https://github.com/exercism/scala/pull/712 they were already not working 2 years ago, and recent new exercises are not even having test-gens.

@ErikSchierboom Does that sound good to you?

**Comment by ErikSchierboom** [comment]
Looks good!

**Comment by grzegorz-bielski** [comment]
@ErikSchierboom I fixed compiler errors in all exercises, please take a moment to review when you have a chance. 🙇 

There are still some warnings, but all tests are passing locally and I believe we can fix them later one by one when porting exercises to a more idiomatic Scala 3 and adding new test-gens.

I also noticed that current tests (bin/test) are _really_ slow. This is because we are starting a new `sbt` session for each exercise. We should consider refactoring it later. Perhaps something similar to [this](https://github.com/exercism/scala/blob/main/project/Build.scala), but for modern sbt. I removed it, since similarly to the test gens, it was already a dead code for some time now, but the idea though is sound.

**Review by ErikSchierboom — COMMENTED** [review]
Great work!

Could you also update the documentation? In particular `docs/INSTALLATION.md`, but there might be other places that need to be updated too.

**Comment by grzegorz-bielski** [comment]
@ErikSchierboom I updated the docs.

On the site there is also `Key Features of Scala`. I think it needs some updates too. The `Lazy computation` point is not really correct. We can take some inspiration from the official scala lang page: https://www.scala-lang.org/.
 Not sure where it can be changed, and what icons are available.

File: docs/ABOUT.md [hunk 1]
```diff
@@ -1,9 +1,12 @@
 # About
 
-Scala combines object-oriented and functional programming in one concise, high-level language. Scala's static types help avoid bugs in complex applications, and its JVM and JavaScript runtimes let you build high-performance systems with easy access to huge ecosystems of libraries.
+A programming language that scales with you: from small scripts to large multiplatform applications.
+
+Scala combines object-oriented and functional programming in one concise, high-level language. Scala's static types help avoid bugs in complex applications, and its JVM, JavaScript and Native runtimes let you build high-performance systems with easy access to huge ecosystems of libraries.
 
 * Features
     * JVM Interoperability - Scala runs on the JVM, so Java and Scala stacks can be mixed for seamless integration.
+    * JS and Native interoperability - Thanks to the [Scala.js](https://www.scala-js.org/) and [Scala Native](https://github.com/scala-native/scala-native)
```
Comment (ErikSchierboom) [inline docs/ABOUT.md:9]:
```suggestion
    * JS and Native interoperability - Scala can run in the browser and on native devides, thanks to [Scala.js](https://www.scala-js.org/) and [Scala Native](https://github.com/scala-native/scala-native)
```

File: docs/ABOUT.md [hunk 2]
```diff
@@ -1,9 +1,12 @@
 # About
 
-Scala combines object-oriented and functional programming in one concise, high-level language. Scala's static types help avoid bugs in complex applications, and its JVM and JavaScript runtimes let you build high-performance systems with easy access to huge ecosystems of libraries.
+A programming language that scales with you: from small scripts to large multiplatform applications.
+
+Scala combines object-oriented and functional programming in one concise, high-level language. Scala's static types help avoid bugs in complex applications, and its JVM, JavaScript and Native runtimes let you build high-performance systems with easy access to huge ecosystems of libraries.
```
Comment (ErikSchierboom) [inline docs/ABOUT.md:5]:
```suggestion
Scala is a a programming language that scales with you: from small scripts to large multiplatform applications.

Scala combines object-oriented and functional programming in one concise, high-level language. Scala's static types help avoid bugs in complex applications, and its JVM, JavaScript and Native runtimes let you build high-performance systems with easy access to huge ecosystems of libraries.
```

**Comment by ErikSchierboom** [comment]
Here we go! 🥳

**Comment by mrybak834** [comment]
Hey, I appreciate the update however this resulted in issues with implicit nulls, seen in exercise [two-fer](https://exercism.org/tracks/scala/exercises/two-fer/).

So I imagine there are other exercise issues related to this. Thanks!

Screenshot 2024-08-14 at 8 47 37 PM
Screenshot 2024-08-14 at 8 51 38 PM

**Comment by mrybak834** [comment]
Thank you for this <3

**Comment by ErikSchierboom** [comment]
> I also noticed that current tests (bin/test) are really slow. This is because we are starting a new sbt session for each exercise. We should consider refactoring it later. Perhaps something similar to [this](https://github.com/exercism/scala/blob/main/project/Build.scala), but for modern sbt. I removed it, since similarly to the test gens, it was already a dead code for some time now, but the idea though is sound.

Improving this would be great!

**Comment by asarkar** [comment]
Hi, thanks for upgrading the Scala track. I started working on the exercises, and see that some of the tests are marked as `pending`. These tests are skipped locally, but occasionally fail when submitted, so, clearly the test runner doesn't care about the `pending` marker. Can you please advise how to run these tests locally so as not to have this discrepancy?

**Comment by ErikSchierboom** [comment]
@asarkar That's documented here: https://exercism.org/docs/tracks/scala/tests

**Pull request merged**